### ISSUE TITLE
Bucket doesn't support Multipart uploads?

### ISSUE THREAD

**Issue by frioux** [rustyconover/net-amazon-s3#16]
I could be misreading this, but at work we have noticed that some uploads fail because they are too big.  This lead me to investigate and we are using [`add_key_filename`](https://metacpan.org/source/RCONOVER/Net-Amazon-S3-0.80/lib/Net/Amazon/S3/Bucket.pm#L75) (thus indirectly [`add_key`](https://metacpan.org/source/RCONOVER/Net-Amazon-S3-0.80/lib/Net/Amazon/S3/Bucket.pm#L41)) to upload files.  This method simply and directly uses a `Net::Amazon::S3::Request::PutObject`, which, as far as I can tell, never attempts to do multipart.  Is it not reasonable to allow some kind of configuration to do multipart based on file size?  Should we be using `Net::Amazon::S3::Client::Object` directly?

**Comment by frioux** [comment]
Maybe more importantly, would you accept a patch to make add_key automatically do multipart, even optionally?

**Comment by TJC** [comment]
+1 -- I've run into issues like this too, and would really like it if multipart uploads were supported inherently. Almost all other S3 libraries default to multipart.

**Comment by ranguard** [comment]
@frioux I'd accept a patch (optionally) https://github.com/rustyconover/net-amazon-s3/pull/4 might also be relevant.

**Comment by ranguard** [comment]
As a side note https://metacpan.org/release/LLAP/Net-Amazon-S3-0.81-TRIAL has been released - testing would be appreciated

**Comment by branislav-zahradnik-gdc** [comment]
I'm working on unifying Bucket and Client::Bucket APIs

**Comment by aravael** [comment]
Hi @branislav-zahradnik-gdc. I'm using version >=0.99, does this library support multipart upload now?

**Comment by happy-barney** [comment]
there is some support in `Net::Amazon::S3::Client` https://metacpan.org/pod/Net::Amazon::S3::Client::Object#initiate_multipart_upload

**Comment by aravael** [comment]
yeah, I'm aware of that. But as @frioux mentioned before the function `add_key_filename` still does not support multipart by default nor even optionally, could you please confirm that?

**Comment by happy-barney** [comment]
`add_key_filename` doesn't support multipart upload. All `add_key` methods behaves like that. Will look into it over weekend.
